### ISSUE TITLE
[BUG/ISSUE] Parallelization issue in GEOS-Chem Classic, probably ISORROPIA

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#693]
@lizziel noticed small numerical differences in GEOS-Chem Classic benchmark output.  I was able to replicate the results with two benchmark runs of 1-hour duration.

Below is the output from GCPy example script `compare_diags.py`.  The script shows the sum of the each data variable in the Ref and Dev run directories, as well as the difference of sums.  In other words, if the difference of sums is zero, both variables are identical.  The Ref code used 6 cores, the Dev code used 4 cores.

```
compare.diags.py: Using configuration file ewl_rst_1h.yml
... Printing totals and differences
... Only showing variables with nonzero differences
Variable               Ref=ewl-Ref              Dev=ewl-Dev              Dev-Ref
Chem_DryDepNitrogen  : 34678556000000.0       | 34678224000000.0       | -331350016.0 
SpeciesRst_HCl       : 0.0002708996           | 0.00027089953          | -5.820766091346741e-11 
SpeciesRst_HNO3      : 0.00026374904          | 0.000263749            | -2.9103830456733704e-11 
SpeciesRst_NIT       : 1.5310257e-05          | 1.531029e-05           | 3.2741809263825417e-11 
SpeciesRst_SALACL    : 2.9291868e-06          | 2.929258e-06           | 7.116796041373163e-11 
```
HCl, HNO3, NIT, and SALACL are species that are modified in ISORROPIA, so it is likely that the differences are entering there.  The difference in DryDepNitrogen is probably an artifact of differing HNO3 concentrations.

I am currently looking into a fix for this.

**Comment by yantosca** [comment]
This issue was caused by 2 separate bugs:

  1. Certain `!$OMP PRIVATE` variables in routine `DO_ISORROPIA` were not zeroed at the start of the loops over I,J,L and I.J,L,NM.  This has now been fixed in commit 547509b484326832104e99aec9f144f82001bc76.
 
  2. In routine `SOIL_DRYDEP` (located in module `GeosCore/get_ndep_mod.F90`), we were saving data from an I,J,L loop into 2-D array `State_Chm%DryDepNitrogen(I,J)`.  This can result in numerical errors when 2 or more threads try to write to this array simultaneously.  To fix this, we have wrapped the assignment statement in an `!$OMP CRITICAL` block, so that only one thread at a time can write to the array.  This was done in commit: 5c70e04350acfd496c05b19e4c06062f1ac2da47.
